### PR TITLE
Add invert and division functions to Expr

### PR DESCRIPTION
**Pull request by mrmr1993**

### CODE REVIEW

File: kimchi/src/circuits/expr.rs [hunk 1]
```diff
@@ -2040,7 +2201,8 @@ impl<F: Neg<Output = F> + Clone + One + Zero + PartialEq> Expr<F> {
                 }
                 res
             }
-        }
+        };
+        Ok(res)
```
Comment (jspada) [inline kimchi/src/circuits/expr.rs:2205]: lgtm

File: kimchi/src/circuits/expr.rs [hunk 2]
```diff
@@ -2206,6 +2368,20 @@ impl<F: Field> Mul<ConstantExpr<F>> for ConstantExpr<F> {
     }
 }
 
+impl<F: Field> Div<ConstantExpr<F>> for ConstantExpr<F> {
+    type Output = ConstantExpr<F>;
+    fn div(self, other: Self) -> Self {
+        use ConstantExpr::{Div, Literal};
+        if other.is_one() {
```
Comment (jspada) [inline kimchi/src/circuits/expr.rs:2375]: Why no `self.is_zero() || other.is_zero()` case?

File: kimchi/src/circuits/expr.rs [hunk 3]
```diff
@@ -2444,8 +2648,10 @@ where
             VanishesOnLast4Rows => "vanishes_on_last_4_rows".to_string(),
             BinOp(Op2::Add, x, y) => format!("({} + {})", x.ocaml(cache), y.ocaml(cache)),
             BinOp(Op2::Mul, x, y) => format!("({} * {})", x.ocaml(cache), y.ocaml(cache)),
+            BinOp(Op2::Div, x, y) => format!("({} / {})", x.ocaml(cache), y.ocaml(cache)),
             BinOp(Op2::Sub, x, y) => format!("({} - {})", x.ocaml(cache), y.ocaml(cache)),
             Pow(x, d) => format!("pow({}, {d})", x.ocaml(cache)),
+            Inv(x) => format!("inv({})", x.ocaml(cache)),
```
Comment (jspada) [inline kimchi/src/circuits/expr.rs:2654]: lgtm!

File: kimchi/src/circuits/expr.rs [hunk 4]
```diff
@@ -2516,8 +2724,10 @@ where
             VanishesOnLast4Rows => "vanishes_on_last_4_rows".to_string(),
             BinOp(Op2::Add, x, y) => format!("({} + {})", x.text(cache), y.text(cache)),
             BinOp(Op2::Mul, x, y) => format!("({} * {})", x.text(cache), y.text(cache)),
+            BinOp(Op2::Div, x, y) => format!("({} / {})", x.text(cache), y.text(cache)),
             BinOp(Op2::Sub, x, y) => format!("({} - {})", x.text(cache), y.text(cache)),
             Pow(x, d) => format!("pow({}, {d})", x.text(cache)),
+            Inv(x) => format!("pow({}, -1)", x.text(cache)),
```
Comment (jspada) [inline kimchi/src/circuits/expr.rs:2730]: why not `"inv({})"` ?

File: kimchi/src/circuits/expr.rs [hunk 5]
```diff
@@ -2400,8 +2602,10 @@ where
                 Alpha => format!("alpha^{}", n),
                 x => format!("{}^{n}", x.text()),
             },
+            Inv(x) => format!("{}^{{-1}}", x.text()),
```
Comment (jspada) [inline kimchi/src/circuits/expr.rs:2605]: suggest: `"inv({})"` -- clearer and easier to search for

File: kimchi/src/circuits/expr.rs [hunk 6]
```diff
@@ -2014,18 +2166,27 @@ impl<F: Neg<Output = F> + Clone + One + Zero + PartialEq> Expr<F> {
...
-                let e1 = e1.monomials(ev);
-                let e2 = e2.monomials(ev);
+                let e1 = e1.monomials(ev)?;
+                let e2 = e2.monomials(ev)?;
+                mul_monomials(&e1, &e2)
+            }
+            BinOp(Op2::Div, e1, e2) => {
+                let e1 = e1.monomials(ev)?;
+                let e2 = if e2.is_constant(ev) {
+                    constant(self.clone())
```
Comment (jspada) [inline kimchi/src/circuits/expr.rs:2176]: Is this right?  Don't you want `e2.monomials(ev)`?

File: kimchi/src/circuits/expr.rs [hunk 7]
```diff
@@ -800,11 +861,14 @@ impl<C> Expr<C> {
             UnnormalizedLagrangeBasis(_) => d1_size,
             Cell(_) => d1_size,
             Square(x) => 2 * x.degree(d1_size),
-            BinOp(Op2::Mul, x, y) => (*x).degree(d1_size) + (*y).degree(d1_size),
+            BinOp(Op2::Mul, x, y) | BinOp(Op2::Div, x, y) => {
+                (*x).degree(d1_size) + (*y).degree(d1_size)
```
Comment (jspada) [inline kimchi/src/circuits/expr.rs:865]: looks good

File: kimchi/src/circuits/expr.rs [hunk 8]
```diff
@@ -563,6 +585,26 @@ impl<C: Zero + One + Neg<Output = C> + PartialEq + Clone> Expr<C> {
...
+                            }
+                        } else if reduce_further2 && c2_reduced.is_one() {
+                            (c1_reduced, false)
+                        } else {
+                            (
+                                BinOp(Op2::Div, Box::new(c1_reduced), Box::new(c2_reduced)),
+                                false,
+                            )
+                        }
+                    }
```
Comment (jspada) [inline kimchi/src/circuits/expr.rs:607]: checked: lgtm!

**Review by jspada — APPROVED** [review]
Looks good to me.  Left some feedback and suggestions.

Spotted one possible mistake.

Also, I think you must add

 * `Div` and an `inverse` method to `pub trait ExprOps<F>:` 
 * Implement `inverse` for `impl<F> ExprOps<F> for Expr<ConstantExpr<F>>`
 * Implement `inverse` for `impl<F: Field> ExprOps<F> for F`

File: kimchi/src/circuits/expr.rs [hunk 9]
```diff
@@ -1263,6 +1327,32 @@ impl<'a, F: FftField> EvalResult<'a, F> {
...
+                Evals { domain, evals }
+            }
+            SubEvals {
+                evals,
+                domain: d,
+                shift: s,
+            } => {
+                let scale = (d as usize) / (res_domain.0 as usize);
+                assert!(scale != 0);
+                let mut res = EvalResult::init(res_domain, |i| {
```
[the remark this answers is not in the document]
Comment (mrmr1993) [reply kimchi/src/circuits/expr.rs:1345]: I sent you a comment by DM, but for posterity: we may be evaluating the constraint over a domain that's smaller than the domain the original evaluations are over. Thus, the scale parameter skips any evaluations that we don't need, creating a vector of only the ones that we do.

This pattern is used a lot, I think we should avoid repeating the documentation everywhere and wait for #886 to be done. Wdyt?

**Comment by mrmr1993** [comment]
This doesn't work as written. Converting to draft, will think about mitigations.

**Comment by mrmr1993** [comment]
This doesn't work as written. Converting to draft, will think about mitigations.

File: kimchi/src/circuits/expr.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (jspada) [reply kimchi/src/circuits/expr.rs:1345]: Yeah, a helper would make the code clearer and less error-prone.  The scale is clear now.  What's the shift parameter for?

[the remark this answers is not in the document]
Comment (mimoo) [reply kimchi/src/circuits/expr.rs:1345]: cc @jspada I wrote some comments there that explains how this works: https://github.com/o1-labs/proof-systems/pull/964

**Comment by mrmr1993** [comment]
Closing, this idea doesn't work.

**Pull request closed without merging**